File: commons/src/main/java/net/xpece/android/widget/XpTimePicker.java
package net.xpece.android.widget;

import android.content.res.ColorStateList;
import android.os.Build;
import android.widget.NumberPicker;
import android.widget.TimePicker;

import java.lang.reflect.Field;

/**
 * Created by Eugen on 06.05.2016.
 */
public final class XpTimePicker {
    private static final Class<?> CLASS_SPINNER_DELEGATE;
    private static final Field FIELD_DELEGATE;

    private static final Field FIELD_HOUR_SPINNER;
    private static final Field FIELD_MINUTE_SPINNER;
    private static final Field FIELD_AM_PM_SPINNER;

    static {
        if (Build.VERSION.SDK_INT >= 11) {
            Field f;
            Class<?> cls = TimePicker.class;
            if (Build.VERSION.SDK_INT < 21) {
                f = null;
                try {
                    f = cls.getDeclaredField("mAmPmSpinner");
                    f.setAccessible(true);
                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                }
                FIELD_AM_PM_SPINNER = f;

                f = null;
                try {
                    f = cls.getDeclaredField("mMinuteSpinner");
                    f.setAccessible(true);
                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                }
                FIELD_MINUTE_SPINNER = f;

                f = null;
                try {
                    f = cls.getDeclaredField("mHourSpinner");
                    f.setAccessible(true);
                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                }
                FIELD_HOUR_SPINNER = f;

                CLASS_SPINNER_DELEGATE = null;
                FIELD_DELEGATE = null;
            } else {
                f = null;
                try {
                    f = cls.getDeclaredField("mDelegate");
                    f.setAccessible(true);
                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                }
                FIELD_DELEGATE = f;

                if (Build.VERSION.SDK_INT == 21) {
                    try {
                        cls = Class.forName("android.widget.TimePickerClockDelegate");
                    } catch (ClassNotFoundException e) {
                        e.printStackTrace();
                    }
                    CLASS_SPINNER_DELEGATE = cls;
                } else {
                    try {
                        cls = Class.forName("android.widget.TimePickerSpinnerDelegate");
                    } catch (ClassNotFoundException e) {
                        e.printStackTrace();
                    }
                    CLASS_SPINNER_DELEGATE = cls;
                }

                f = null;
                try {
                    f = cls.getDeclaredField("mAmPmSpinner");
                    f.setAccessible(true);
                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                }
                FIELD_AM_PM_SPINNER = f;

                f = null;
                try {
                    f = cls.getDeclaredField("mMinuteSpinner");
                    f.setAccessible(true);
                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                }
                FIELD_MINUTE_SPINNER = f;

                f = null;
                try {
                    f = cls.getDeclaredField("mHourSpinner");
                    f.setAccessible(true);
                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                }
                FIELD_HOUR_SPINNER = f;
            }
        } else {
            FIELD_AM_PM_SPINNER = null;
            FIELD_HOUR_SPINNER = null;
            FIELD_MINUTE_SPINNER = null;
            FIELD_DELEGATE = null;
            CLASS_SPINNER_DELEGATE = null;
        }
    }

    private XpTimePicker() {}

    private static NumberPicker getAmPmSpinner(Object picker) {
        try {
            return (NumberPicker) FIELD_AM_PM_SPINNER.get(picker);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            return null;
        }
    }

    private static NumberPicker getMinuteSpinner(Object picker) {
        try {
            return (NumberPicker) FIELD_MINUTE_SPINNER.get(picker);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            return null;
        }
    }

    private static NumberPicker getHourSpinner(Object picker) {
        try {
            return (NumberPicker) FIELD_HOUR_SPINNER.get(picker);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            return null;
        }
    }

    private static Object getDelegate(TimePicker picker) {
        try {
            return FIELD_DELEGATE.get(picker);
        } catch (IllegalAccessException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static void setSelectionDividerTint(TimePicker picker, ColorStateList color) {
        if (Build.VERSION.SDK_INT < 11) return;

        if (Build.VERSION.SDK_INT < 21) {
            NumberPicker np;
            np = getAmPmSpinner(picker);
            if (np != null) {
                XpNumberPicker.setSelectionDividerTintInt(np, color);
            }
            np = getMinuteSpinner(picker);
            if (np != null) {
                XpNumberPicker.setSelectionDividerTintInt(np, color);
            }
            np = getHourSpinner(picker);
            if (np != null) {
                XpNumberPicker.setSelectionDividerTintInt(np, color);
            }
        } else {
            Object delegate = getDelegate(picker);
            if (delegate != null && delegate.getClass() == CLASS_SPINNER_DELEGATE) {
                NumberPicker np;
                np = getAmPmSpinner(delegate);
                if (np != null) {
                    XpNumberPicker.setSelectionDividerTintInt(np, color);
                }
                np = getMinuteSpinner(delegate);
                if (np != null) {
                    XpNumberPicker.setSelectionDividerTintInt(np, color);
                }
                np = getHourSpinner(delegate);
                if (np != null) {
                    XpNumberPicker.setSelectionDividerTintInt(np, color);
                }
            }
        }
    }
}

File: commons/src/main/java/net/xpece/android/widget/XpTimePicker.kt
@file:JvmName("XpTimePicker")

package net.xpece.android.widget

import android.content.res.ColorStateList
import android.os.Build
import android.util.Log
import android.widget.NumberPicker
import android.widget.TimePicker
import java.lang.reflect.Field

private object TimePickerReflection {

    internal val delegateClass: Class<out Any>?
    internal val delegateField: Field?

    internal val minuteSpinnerField: Field?
    internal val hourSpinnerField: Field?
    internal val amPmSpinnerField: Field?

    init {
        val implClass: Class<out Any>?
        if (Build.VERSION.SDK_INT < 21) {
            delegateField = null
            delegateClass = null
            implClass = TimePicker::class.java
        } else {
            delegateField = TimePicker::class.java.getOptionalPrivateField("mDelegate")
            val delegateName = if (Build.VERSION.SDK_INT == 21) {
                // On Android 5.0 the two delegate class names are switched.
                "android.widget.TimePickerClockDelegate"
            } else {
                "android.widget.TimePickerSpinnerDelegate"
            }
            delegateClass = try {
                Class.forName(delegateName)
            } catch (ex: ClassNotFoundException) {
                Log.w("XpTimePicker", "Couldn't fetch $delegateName class.", ex)
                null
            }
            implClass = delegateClass
        }
        if (implClass != null) {
            minuteSpinnerField = implClass.getOptionalPrivateField("mMinuteSpinner")
            hourSpinnerField = implClass.getOptionalPrivateField("mHourSpinner")
            amPmSpinnerField = implClass.getOptionalPrivateField("mAmPmSpinner")
        } else {
            minuteSpinnerField = null
            hourSpinnerField = null
            amPmSpinnerField = null
        }
    }

    private fun <T> Class<T>.getOptionalPrivateField(name: String): Field? = try {
        getDeclaredField(name)
                .apply { isAccessible = true }
    } catch (ex: NoSuchFieldException) {
        Log.w("XpTimePicker", "Couldn't fetch $simpleName.$name field.", ex)
        null
    }
}

private fun TimePicker.getDelegate(): Any? = if (TimePickerReflection.delegateField != null) {
    TimePickerReflection.delegateField.get(this)
} else {
    Log.w("XpTimePicker", "Couldn't get TimePicker.mDelegate field.")
    null
}

private fun Any.getDaySpinner(): NumberPicker? = if (TimePickerReflection.minuteSpinnerField != null) {
    TimePickerReflection.minuteSpinnerField.get(this) as NumberPicker?
} else {
    Log.w("XpTimePicker", "Couldn't get TimePickerSpinnerDelegate.mMinuteSpinner field.")
    null
}

private fun Any.getMonthSpinner(): NumberPicker? = if (TimePickerReflection.hourSpinnerField != null) {
    TimePickerReflection.hourSpinnerField.get(this) as NumberPicker?
} else {
    Log.w("XpTimePicker", "Couldn't get TimePickerSpinnerDelegate.mHourSpinner field.")
    null
}

private fun Any.getYearSpinner(): NumberPicker? = if (TimePickerReflection.amPmSpinnerField != null) {
    TimePickerReflection.amPmSpinnerField.get(this) as NumberPicker?
} else {
    Log.w("XpTimePicker", "Couldn't get TimePickerSpinnerDelegate.mAmPmSpinner field.")
    null
}

fun TimePicker.setSelectionDividerTint(color: ColorStateList?) {
    val impl = if (Build.VERSION.SDK_INT < 21) {
        this
    } else {
        getDelegate()?.takeIf { javaClass == TimePickerReflection.delegateClass }
    }
    impl?.apply {
        getYearSpinner()?.setSelectionDividerTint(color)
        getMonthSpinner()?.setSelectionDividerTint(color)
        getDaySpinner()?.setSelectionDividerTint(color)
    }
}
